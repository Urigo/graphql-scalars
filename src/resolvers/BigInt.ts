// Based on https://github.com/stems/graphql-bigint/

import { Kind, GraphQLScalarType } from 'graphql';

export default function (name = 'BigInt') {
    return new GraphQLScalarType({
        name,
        description:
            'The `BigInt` scalar type represents non-fractional signed whole numeric values.',
        serialize: BigInt,
        parseValue: BigInt,
        parseLiteral(ast) {
            if (ast.kind === Kind.INT || ast.kind === Kind.FLOAT || ast.kind === Kind.STRING) {
                return BigInt(ast.value);
            }
            return null;
        }
    });
}
