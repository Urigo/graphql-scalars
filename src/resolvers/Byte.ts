import {
  ASTNode,
  GraphQLScalarType,
  IntValueNode,
  Kind,
  ObjectValueNode,
  print,
} from 'graphql';

const base64Validator = /^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/;
const hexValidator = /(0x|0X)?[a-fA-F0-9]+$/;

type BufferJson = { type: 'Buffer'; data: number[] };

function validate(value: Buffer | string | BufferJson) {
  if (typeof value !== 'string' && !(value instanceof Buffer)) {
    throw new TypeError(
      `Value is not an instance of Buffer: ${JSON.stringify(value)}`,
    );
  }
  if (typeof value === 'string') {
    const isBase64 = base64Validator.test(value);
    const isHex = hexValidator.test(value);
    if (!isBase64 && !isHex) {
      throw new TypeError(
        `Value is not a valid base64 or hex encoded string: ${JSON.stringify(
          value,
        )}`,
      );
    }
    return Buffer.from(value, isHex ? 'hex' : 'base64');
  }

  return value;
}

function parseObject(ast: ObjectValueNode) {
  const key = ast.fields[0].value;
  const value = ast.fields[1].value;
  if (
    ast.fields.length === 2 &&
    key.kind === Kind.STRING &&
    key.value === 'Buffer' &&
    value.kind === Kind.LIST
  ) {
    return Buffer.from(
      value.values.map((astValue: IntValueNode) => parseInt(astValue.value)),
    );
  }
  throw new TypeError(
    `Value is not a JSON representation of Buffer: ${print(ast)}`,
  );
}

export default new GraphQLScalarType({
  name: 'Byte',
  description: 'The `Byte` scalar type represents byte value as a Buffer',
  serialize: validate,
  parseValue: validate,
  parseLiteral(ast: ASTNode) {
    switch (ast.kind) {
      case Kind.STRING:
        return validate(ast.value);
      case Kind.OBJECT:
        return parseObject(ast);
      default:
        throw new TypeError(
          `Can only parse base64 or hex encoded strings as Byte, but got a: ${ast.kind}`,
        );
    }
  },
});
