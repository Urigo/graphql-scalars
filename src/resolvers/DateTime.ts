import { Kind, GraphQLError, GraphQLScalarType } from 'graphql';

export default new GraphQLScalarType({
  name: 'DateTime',

  description: 'Use JavaScript Date object for date/time fields.',

  serialize(value) {
    let v = value;

    if (!(v instanceof Date) && typeof v !== 'string' && typeof v !== 'number') {
      throw new TypeError(
        `Value is not an instance of Date, Date string or number: ${JSON.stringify(v)}`,
      );
    }

    if (typeof v === 'string') {
      v = new Date();

      v.setTime(Date.parse(value));
    } else if (typeof v === 'number') {
      v = new Date(v);
    }

    // eslint-disable-next-line no-restricted-globals
    if (Number.isNaN(v.getTime())) {
      throw new TypeError(`Value is not a valid Date: ${JSON.stringify(v)}`);
    }

    return v.toJSON();
  },

  parseValue(value) {
    const date = new Date(value);

    // eslint-disable-next-line no-restricted-globals
    if (Number.isNaN(date.getTime())) {
      throw new TypeError(`Value is not a valid Date: ${value}`);
    }

    return date;
  },

  parseLiteral(ast) {
    if (ast.kind !== Kind.STRING && ast.kind !== Kind.INT) {
      throw new GraphQLError(
        `Can only parse strings & integers to dates but got a: ${ast.kind}`,
      );
    }

    const result = new Date(ast.kind === Kind.INT ? Number(ast.value) : ast.value);

    // eslint-disable-next-line no-restricted-globals
    if (Number.isNaN(result.getTime())) {
      throw new GraphQLError(`Value is not a valid Date: ${ast.value}`);
    }

    return result;
  },
});
