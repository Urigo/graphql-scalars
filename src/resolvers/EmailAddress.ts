import { Kind, GraphQLError, GraphQLScalarType } from 'graphql';

/* eslint-disable no-useless-escape */
const EMAIL_ADDRESS_REGEX = new RegExp(
  /^[a-zA-Z0-9.!#$%&â€™*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/,
);
/* eslint-enable */

export default new GraphQLScalarType({
  name: 'EmailAddress',

  description:
    'A field whose value conforms to the standard internet email address format as specified in RFC822: https://www.w3.org/Protocols/rfc822/.',

  serialize(value) {
    if (typeof value !== 'string') {
      throw new TypeError(`Value is not string: ${value}`);
    }

    if (!EMAIL_ADDRESS_REGEX.test(value)) {
      throw new TypeError(`Value is not a valid email address: ${value}`);
    }

    return value;
  },

  parseValue(value) {
    if (typeof value !== 'string') {
      throw new TypeError('Value is not string');
    }

    if (!EMAIL_ADDRESS_REGEX.test(value)) {
      throw new TypeError(`Value is not a valid email address: ${value}`);
    }

    return value;
  },

  parseLiteral(ast) {
    if (ast.kind !== Kind.STRING) {
      throw new GraphQLError(
        `Can only validate strings as email addresses but got a: ${ast.kind}`,
      );
    }

    if (!EMAIL_ADDRESS_REGEX.test(ast.value)) {
      throw new TypeError(`Value is not a valid email address: ${ast.value}`);
    }

    return ast.value;
  },
});
