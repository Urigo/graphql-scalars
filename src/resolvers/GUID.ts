import { Kind, GraphQLError, GraphQLScalarType } from 'graphql';

const GUID_REGEX = /^[0-9a-f]{8}-?[0-9a-f]{4}-?[1-5][0-9a-f]{3}-?[89ab][0-9a-f]{3}-?[0-9a-f]{12}$/i;

const validate = (value: any) => {
  if (typeof value !== 'string') {
    throw new TypeError(`Value is not string: ${value}`);
  }

  if (!(GUID_REGEX.test(value))) {
    throw new TypeError(`Value is not a valid GUID: ${value}`);
  }

  return value;
};

export default new GraphQLScalarType({
  name: `GUID`,

  description: `A field whose value is a generic Globally Unique Identifier: https://en.wikipedia.org/wiki/Universally_unique_identifier.`,

  serialize(value) {
    return validate(value);
  },

  parseValue(value) {
    return validate(value);
  },

  parseLiteral(ast) {
    if (ast.kind !== Kind.STRING) {
      throw new GraphQLError(`Can only validate strings as GUIDs but got a: ${ast.kind}`);
    }

    return validate(ast.value);
  }
});
