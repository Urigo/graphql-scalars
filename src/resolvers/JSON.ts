import { Kind, GraphQLScalarType, ValueNode, ObjectValueNode } from 'graphql';

function identity<T>(value: T): T {
    return value;
}

function ensureObject(value: any): value is object {
    if (typeof value !== 'object' || value === null || Array.isArray(value)) {
        throw new TypeError(
            `JSONObject cannot represent non-object value: ${value}`,
        );
    }

    return value;
}

function parseObject(ast: ObjectValueNode, variables: any): any {
    const value = Object.create(null);
    ast.fields.forEach(field => {
        // eslint-disable-next-line no-use-before-define
        value[field.name.value] = parseLiteral(field.value, variables);
    });

    return value;
}

function parseLiteral(ast: ValueNode, variables: any): any {
    switch (ast.kind) {
        case Kind.STRING:
        case Kind.BOOLEAN:
            return ast.value;
        case Kind.INT:
        case Kind.FLOAT:
            return parseFloat(ast.value);
        case Kind.OBJECT:
            return parseObject(ast, variables);
        case Kind.LIST:
            return ast.values.map(n => parseLiteral(n, variables));
        case Kind.NULL:
            return null;
        case Kind.VARIABLE: {
            const name = ast.name.value;
            return variables ? variables[name] : undefined;
        }
    }
}

// This named export is intended for users of CommonJS. Users of ES modules
// should instead use the default export.
export const JSON = new GraphQLScalarType({
    name: 'JSON',
    description:
        'The `JSON` scalar type represents JSON values as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf).',
    serialize: identity,
    parseValue: identity,
    parseLiteral,
});

export const JSONObject = new GraphQLScalarType({
    name: 'JSONObject',
    description:
        'The `JSONObject` scalar type represents JSON objects as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf).',
    serialize: ensureObject,
    parseValue: ensureObject,
    parseLiteral: parseObject,
});
