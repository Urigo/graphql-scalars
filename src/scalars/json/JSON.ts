// This named export is intended for users of CommonJS. Users of ES modules

import { GraphQLScalarType, GraphQLScalarTypeConfig } from 'graphql';
import { identity, parseLiteral } from './utils';

export const GraphQLJSONConfig: GraphQLScalarTypeConfig<any, any> =
  /*#__PURE__*/ {
    name: 'JSON',
    description:
      'The `JSON` scalar type represents JSON values as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf).',
    serialize: identity,
    parseValue: identity,
    parseLiteral,
    specifiedByURL:
      'http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf',
    extensions: {
      codegenScalarType: 'any',
    },
  };

export const GraphQLJSON = /*#__PURE__*/ new GraphQLScalarType(
  GraphQLJSONConfig,
);
