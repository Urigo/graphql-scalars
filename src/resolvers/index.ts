import DateTime from './DateTime';
import NonPositiveInt from './NonPositiveInt';
import PositiveInt from './PositiveInt';
import NonNegativeIntFactory from './NonNegativeInt';
import NegativeInt from './NegativeInt';
import NonPositiveFloat from './NonPositiveFloat';
import PositiveFloat from './PositiveFloat';
import NonNegativeFloatFactory from './NonNegativeFloat';
import NegativeFloat from './NegativeFloat';
import EmailAddress from './EmailAddress';
import URL from './URL';
import PhoneNumber from './PhoneNumber';
import PostalCode from './PostalCode';
import BigIntFactory from './BigInt';
import GUID from './GUID';
import Hexadecimal from './Hexadecimal';
import HexColorCode from './HexColorCode';
import HSL from './HSL';
import HSLA from './HSLA';
import IPv4 from './IPv4';
import IPv6 from './IPv6';
import ISBN from './ISBN';
import MAC from './MAC';
import Port from './Port';
import RGB from './RGB';
import RGBA from './RGBA';
import USCurrency from './USCurrency';
import { JSON, JSONObject } from './JSON';
import IBAN from './IBAN';

const BigIntResolver = BigIntFactory('BigInt');
const LongResolver = BigIntFactory('Long');

const NonNegativeIntResolver = NonNegativeIntFactory('NonNegativeInt');
const UnsignedIntResolver = NonNegativeIntFactory('UnsignedInt');

const NonNegativeFloatResolver = NonNegativeFloatFactory('NonNegativeFloat');
const UnsignedFloatResolver = NonNegativeFloatFactory('UnsignedFloat');

export {
    DateTime,
    NonPositiveInt,
    PositiveInt,
    NonNegativeIntResolver as NonNegativeInt,
    UnsignedIntResolver as UnsignedInt,
    NegativeInt,
    NonPositiveFloat,
    PositiveFloat,
    NonNegativeFloatResolver as NonNegativeFloat,
    UnsignedFloatResolver as UnsignedFloat,
    NegativeFloat,
    EmailAddress,
    URL,
    PhoneNumber,
    PostalCode,
    BigIntResolver as BigInt,
    LongResolver as Long,
    GUID,
    Hexadecimal,
    HexColorCode,
    HSL,
    HSLA,
    IPv4,
    IPv6,
    ISBN,
    MAC,
    Port,
    RGB,
    RGBA,
    USCurrency,
    JSON,
    JSONObject,
    IBAN
};
