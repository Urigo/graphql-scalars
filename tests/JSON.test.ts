import {
    graphql,
    GraphQLInt,
    GraphQLObjectType,
    GraphQLSchema,
} from 'graphql';

// eslint-disable-next-line import/no-named-as-default
import { JSON as GraphQLJSON, JSONObject as GraphQLJSONObject } from '../src/resolvers/JSON';

const FIXTURE = {
    string: 'string',
    int: 3,
    float: 3.14,
    true: true,
    false: false,
    null: null,
    object: {
        string: 'string',
        int: 3,
        float: 3.14,
        true: true,
        false: false,
        null: null,
    },
    array: ['string', 3, 3.14, true, false, null],
};

function createSchema(type) {
    return new GraphQLSchema({
        query: new GraphQLObjectType({
            name: 'Query',
            fields: {
                value: {
                    type,
                    args: {
                        arg: { type },
                    },
                    resolve: (obj, { arg }) => arg,
                },
                rootValue: {
                    type,
                    resolve: obj => obj,
                },
            },
        }),
        types: [GraphQLInt],
    });
}

describe('GraphQLJSON', () => {
    let schema;

    beforeEach(() => {
        schema = createSchema(GraphQLJSON);
    });

    describe('serialize', () => {
        it('should support serialization', () =>
            graphql(
                schema,
          /* GraphQL */ `
            query {
              rootValue
            }
          `,
                FIXTURE,
            ).then(({ data, errors }) => {
                expect(data.rootValue).toEqual(FIXTURE);
                expect(errors).toBeUndefined();
            }));
    });

    describe('parseValue', () => {
        it('should support parsing values', () =>
            graphql(
                schema,
          /* GraphQL */ `
            query($arg: JSON!) {
              value(arg: $arg)
            }
          `,
                null,
                null,
                {
                    arg: FIXTURE,
                },
            ).then(({ data, errors }) => {
                expect(data.value).toEqual(FIXTURE);
                expect(errors).toBeUndefined();
            }));
    });

    describe('parseLiteral', () => {
        it('should support parsing literals', () =>
            graphql(
                schema,
          /* GraphQL */ `
            query($intValue: Int = 3) {
              value(
                arg: {
                  string: "string"
                  int: $intValue
                  float: 3.14
                  true: true
                  false: false
                  null: null
                  object: {
                    string: "string"
                    int: $intValue
                    float: 3.14
                    true: true
                    false: false
                    null: null
                  }
                  array: ["string", $intValue, 3.14, true, false, null]
                }
              )
            }
          `,
            ).then(({ data, errors }) => {
                expect(data.value).toEqual(FIXTURE);
                expect(errors).toBeUndefined();
            }));

        it('should handle null literal', () =>
            graphql(
                schema,
          /* GraphQL */ `
            {
              value(arg: null)
            }
          `,
            ).then(({ data, errors }) => {
                expect(data).toEqual({
                    value: null,
                });
                expect(errors).toBeUndefined();
            }));

        it('should handle list literal', () =>
            graphql(
                schema,
          /* GraphQL */ `
            {
              value(arg: [])
            }
          `,
            ).then(({ data, errors }) => {
                expect(data).toEqual({
                    value: [],
                });
                expect(errors).toBeUndefined();
            }));

        it('should reject invalid literal', () =>
            graphql(
                schema,
          /* GraphQL */ `
            {
              value(arg: INVALID)
            }
          `,
            ).then(({ data, errors }) => {
                expect(data).toBeUndefined();
                expect(errors).toBeDefined();
            }));
    });
});

describe('GraphQLJSONObject', () => {
    let schema;

    beforeEach(() => {
        schema = createSchema(GraphQLJSONObject);
    });

    describe('serialize', () => {
        it('should support serialization', () =>
            graphql(
                schema,
          /* GraphQL */ `
            query {
              rootValue
            }
          `,
                FIXTURE,
            ).then(({ data, errors }) => {
                expect(data.rootValue).toEqual(FIXTURE);
                expect(errors).toBeUndefined();
            }));

        it('should reject string value', () =>
            graphql(
                schema,
          /* GraphQL */ `
            query {
              rootValue
            }
          `,
                'foo',
            ).then(({ data, errors }) => {
                expect(data.rootValue).toBeNull();
                expect(errors).toBeDefined();
            }));

        it('should reject array value', () =>
            graphql(
                schema,
          /* GraphQL */ `
            query {
              rootValue
            }
          `,
                [],
            ).then(({ data, errors }) => {
                expect(data.rootValue).toBeNull();
                expect(errors).toBeDefined();
            }));
    });

    describe('parseValue', () => {
        it('should support parsing values', () =>
            graphql(
                schema,
          /* GraphQL */ `
            query($arg: JSONObject!) {
              value(arg: $arg)
            }
          `,
                null,
                null,
                {
                    arg: FIXTURE,
                },
            ).then(({ data, errors }) => {
                expect(data.value).toEqual(FIXTURE);
                expect(errors).toBeUndefined();
            }));

        it('should reject string value', () =>
            graphql(
                schema,
          /* GraphQL */ `
            query($arg: JSONObject!) {
              value(arg: $arg)
            }
          `,
                null,
                null,
                {
                    arg: 'foo',
                },
            ).then(({ data, errors }) => {
                expect(data).toBeUndefined();
                expect(errors).toBeDefined();
            }));

        it('should reject array value', () =>
            graphql(
                schema,
          /* GraphQL */ `
            query($arg: JSONObject!) {
              value(arg: $arg)
            }
          `,
                null,
                null,
                {
                    arg: [],
                },
            ).then(({ data, errors }) => {
                expect(data).toBeUndefined();
                expect(errors).toBeDefined();
            }));
    });

    describe('parseLiteral', () => {
        it('should support parsing literals', () =>
            graphql(
                schema,
          /* GraphQL */ `
            query($intValue: Int = 3) {
              value(
                arg: {
                  string: "string"
                  int: $intValue
                  float: 3.14
                  true: true
                  false: false
                  null: null
                  object: {
                    string: "string"
                    int: $intValue
                    float: 3.14
                    true: true
                    false: false
                    null: null
                  }
                  array: ["string", $intValue, 3.14, true, false, null]
                }
              )
            }
          `,
            ).then(({ data, errors }) => {
                expect(data.value).toEqual(FIXTURE);
                expect(errors).toBeUndefined();
            }));

        it('should reject string literal', () =>
            graphql(
                schema,
          /* GraphQL */ `
            {
              value(arg: "foo")
            }
          `,
            ).then(({ data, errors }) => {
                expect(data).toBeUndefined();
                expect(errors).toBeDefined();
            }));

        it('should reject array literal', () =>
            graphql(
                schema,
          /* GraphQL */ `
            {
              value(arg: [])
            }
          `,
            ).then(({ data, errors }) => {
                expect(data).toBeUndefined();
                expect(errors).toBeDefined();
            }));
    });
});
