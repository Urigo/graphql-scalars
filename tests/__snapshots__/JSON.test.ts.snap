// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQLJSONObject parseLiteral should reject string literal 1`] = `
[
  [GraphQLError: JSONObject cannot represent non-object value: "foo"],
]
`;
